src/algorithm/bluestein.rs

```diff
@@ -0,0 +1,207 @@
+            }
+        } {
+            let twiddle = calculate_twiddle(index, len)*scale;
+            if inverse {
+                *tw = twiddle;
+            }
+            else {
+                *tw = twiddle.conj();
+            }
+        } 
+    }
+    fft.process(&mut scratch, &mut twiddles[..]);
+}
+
+
+/// Calculate the "x" twiddles.
+fn calculate_x_twiddles<T: FFTnum>(
+    len: usize,
+    twiddles: &mut [Complex<T>],
+    inverse: bool,
+) {
+    if inverse {
+        for (i, tw) in twiddles.iter_mut().enumerate() {
+            *tw = calculate_twiddle(-(i as f64).powi(2), len);
+        }
+    }
+    else {
+        for (i, tw) in twiddles.iter_mut().enumerate() {
+            *tw = calculate_twiddle(-(i as f64).powi(2), len).conj();
+        }
+    }
+}
+
+
+impl<T: FFTnum > Bluesteins<T> {
+    /// Creates a FFT instance which will process inputs/outputs of size `len`. `inner_fft_fw.len()` and `inner_fft_inv.len()` must be the nearest 
+    /// power of two that is equal to or larger than `2 * len - 1`.
+    ///
+    /// Note that this constructor is quite expensive to run; This algorithm must run a FFT within the
+    /// constructor. 
+    pub fn new(len: usize, inner_fft_fw: Arc<FFT<T>>, inner_fft_inv: Arc<FFT<T>>, inverse: bool) -> Self {
+        let inner_fft_len = (2 * len - 1).checked_next_power_of_two().unwrap();
+        assert_eq!(inner_fft_len, inner_fft_fw.len(), "For Bluesteins algorithm, inner_fft_fw.len() must be a power of to larger than or equal to 2*self.len() - 1. Expected {}, got {}", inner_fft_len, inner_fft_fw.len());
+        assert_eq!(inner_fft_len, inner_fft_inv.len(), "For Bluesteins algorithm, inner_fft_inv.len() must be a power of to larger than or equal to 2*self.len() - 1. Expected {}, got {}", inner_fft_len, inner_fft_inv.len());
+
+        let mut w_twiddles = vec![Complex::zero(); inner_fft_len];
+        let mut x_twiddles = vec![Complex::zero(); len];
+        calculate_w_twiddles(len, &inner_fft_fw, &mut w_twiddles, inverse);
+        calculate_x_twiddles(len, &mut x_twiddles, inverse);
+        Self {
+            len,
+            inner_fft_fw,
+            inner_fft_inv,
+            w_twiddles: w_twiddles.into_boxed_slice(),
+            x_twiddles: x_twiddles.into_boxed_slice(),
+            inverse,
+        }
+    }
+
+    fn perform_fft(&self, input: &mut [Complex<T>], output: &mut [Complex<T>]) {
+        assert_eq!(self.len(), input.len());
+
+        let mut scratch_a = vec![Complex::zero(); self.inner_fft_fw.len()];
+        let mut scratch_b = vec![Complex::zero(); self.inner_fft_fw.len()];
+        for (w, (x, i)) in scratch_a.iter_mut().zip(self.x_twiddles.iter().zip(input.iter())) {
+            *w = x * i;
+        }
+        self.inner_fft_fw.process(&mut scratch_a, &mut scratch_b);
+        for (w, wi) in scratch_b.iter_mut().zip(self.w_twiddles.iter()) {
+            *w = *w * wi;
+        }
+        self.inner_fft_inv.process(&mut scratch_b, &mut scratch_a);
```

Comment:
I have an optimization to recommend: You can compute an inverse FFT by:
 1. Conjugating the inputs
 1. Computing a forward FFT on the conjugated data
 1. Conjugating the output of the forward FFT

Conjugating will be very easy, since you're already looping over both the inputs and outputs, and it means this algorithm only needs to store a forward FFT, not both a forward and inverse.

Reply:
It will also mean we don't need to take "inverse" in the constructor either, because we can grab it from the inner FFT.

Reply:
Good idea! I will do this and update the PR.